proposals/ytt/004-schema-validation/README.md

```diff
@@ -0,0 +1,1523 @@
+
+Note: if an author wishes to provide a custom violation message, they would use the corresponding built-in rule as a custom rule.
+
+#### @assert/validate-string-defaults
+
+> **TODO:**
+> - [ ] firm-up name of this annotation: is it a bug or a feature that it is type-specific?
+
+Defines what a valid value is for all descendants which hold a "string" value.
+
+```
+@assert/validate-string-defaults [rule0, rule1...] [,<named-rules>] [,when=] [,when_null_skip=]
+```
+
+See also: [Consideration: Setting validation defaults for strings in a schema overlay](#consideration-setting-validation-defaults-for-strings-in-a-schema-overlay).
+
+#### @ytt:assert.validate()
+
+Programmatically asserts that a given value is valid based on the set of supplied rules.
+
+```python
+assert.validate(key, value, rule0, rule1, ... [,<named-rules>] [,when=] [,when_null_skip=])
+```
+where:
+- `key` (`string`) — the name of the value being validated.
+- `value` (any) — the value being validated.
+
+This function is the programmatic equivalent of [@assert/validate](#assertvalidate).
+
+#### @ytt:assert.is_valid()
+
+Asserts any validations present on a Node (and descendents).
+
+```python
+assert.is_valid(node)
+```
+
+- if all validation rules are satisfied, nothing happens.
+- if any validation rule is violated, `assert.fail()`s with the corresponding validation message(s).
+
+
+#### Included Assertion Functions
+
+- the functions in this section are technically higher-order functions that _produce_ assertions:
+    ```python
+    def assert.min(x):
+      return lambda v: assert.fail("at least {}", x) if v >= x else None
+    end
+    ```
+    of which the return value _is_ an assertion.
+- violation messages should be consistent across these functions.
+- these functions should attempt to support as many types as possible:
+
+##### @ytt:assert.contains()
+
+##### @ytt:assert.ends_with()
+
+##### @ytt:assert.format()
+
+Supported formats:
+- "quantity" — https://kubernetes.io/docs/reference/glossary/?all=true#term-quantity
+- Additional keywords will be added as needs are clear. Likely candidates come from common domain-specific formats:
+[JSON Schema Validation: §7 Vocabularies for Semantic Content With "format"](https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.7).
+
+##### @ytt:assert.len()
+
+The length of the value is _exactly_ that of the given length.
+
+##### @ytt:assert.matches()
+
+##### @ytt:assert.max()
+
+The value is _at most_ (inclusive) of the given minimum.
+
+##### @ytt:assert.max_len()
+
+The length of the value is _at most_ (inclusive) that of the given minimum.
+
+
+##### @ytt:assert.min()
+
+The value is _at least_ (inclusive) of the given minimum.
+
+##### @ytt:assert.min_len()
+
+The length of the value is _at least_ (inclusive) that of the given minimum.
+
+##### @ytt:assert.not_null()
+
+When present, this validator is always checked first.
+
+##### @ytt:assert.one_not_null()
+
+##### @ytt:assert.one_of()
```

Comment:
Q: is this confusingly similar to OpenAPI v3's `oneOf` validation keyword? _That_ keyword is the means to express a proper union, where as this assertion is about an enumeration.

```suggestion
##### @ytt:assert.enum()
```

Reply:
Isn't the intent behind this validation to check that the user selected one of the following values? If so I believe `one_of` conveys that intent.